File: src/main/java/algo/ds/Trie.java
package algo.ds;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;


public class Trie {

    private static class Node {
        boolean isWord;
        final Node[] childs = new Node['z' - 'a' + 1];
    }


    private final Node root = new Node();


    public void insert(String word) {
        Node node = root;
        for (int i = 0; i < word.length(); ++ i) {
            int c = word.charAt(i) - 'a';
            if (node.childs[c] == null) {
                node.childs[c] = new Node();
            }
            node = node.childs[c];
        }
        node.isWord = true;
    }


    public boolean search(String word) {
        Node node = getNode(word);
        return node != null && node.isWord;
    }



    public boolean startsWith(String prefix) {
        return getNode(prefix) != null;
    }


    private Node getNode(String word) {
        Node node = root;
        for (int i = 0; i < word.length(); ++ i) {
            int c = word.charAt(i) - 'a';
            node = node.childs[c];
            if (node == null) {
                return null;
            }
        }
        return node;
    }


    public List<Node> getWords(Node parent) {
        List<Node> result = new ArrayList<>();

        Deque<Node> deque = new ArrayDeque<>();
        deque.addLast(parent);

        while (!deque.isEmpty()) {
            int size = deque.size();

            while (size -- > 0) {
                Node node = deque.removeFirst();
                if (node.isWord) {
                    result.add(node);
                }

                for (Node n : node.childs) {
                    if (n != null) {
                        deque.addLast(n);
                    }
                }
            }
        }

        return result;
    }

}

File: python/ds/trie.py
import dataclasses


@dataclasses.dataclass
class Node:
    is_word: bool
    childs: dataclasses.field(default_factory=list)

    def __init__(self):
        self.is_word = False
        self.childs = [None] * 26


@dataclasses.dataclass
class Trie:
    root: Node = Node()

    def insert(self, word: str):
        node = self.root
        for c in word:
            i = ord(c) - ord('a')
            if node.childs[i] is None:
                node.childs[i] = Node()
            node = node.childs[i]
        node.is_word = True


    def get_node(self, word: str) -> Node:
        node = self.root
        for c in word:
            i = ord(c) - ord('a')
            node = node.childs[i]
            if node is None:
                return None
        return node


if __name__ == '__main__':
    _trie = Trie()
    assert _trie.get_node('unkn') is None
    _trie.insert('word')
    _node = _trie.get_node('wor')
    assert _node is not None and not _node.is_word
    _node = _trie.get_node('word')
    assert _node is not None and _node.is_word
